snuba/query/query_processor.py

```diff
@@ -40,11 +46,59 @@ class ExtensionQueryProcessor(QueryProcessor[ExtensionData]):
     """
 
     @abstractmethod
-    def process_query(self, query: Query, extension_data: ExtensionData) -> None:
+    def process_query(
+            self, query: Query,
+            extension_data: ExtensionData,
+            request_settings: Mapping[str, bool],
+            query_hints: Mapping[str, Any],
+            stats: Mapping[str, Any],
+    ) -> None:
         raise NotImplementedError
 
 
 class DummyExtensionProcessor(ExtensionQueryProcessor):
 
-    def process_query(self, query: Query, extension_data: ExtensionData) -> None:
+    def process_query(
+            self,
+            query: Query,
+            extension_data: ExtensionData,
+            request_settings: Mapping[str, bool],
+            query_hints: Mapping[str, Any],
+            stats: Mapping[str, Any],
+    ) -> None:
         return query
+
+
+class ProjectExtensionProcessor(ExtensionQueryProcessor):
+
+    def process_query(
+            self,
+            query: Query,
+            extension_data: ExtensionData,
+            request_settings: Mapping[str, bool],
+            query_hints: Mapping[str, Any],
+            stats: Mapping[str, Any],
+    ) -> None:
+        # NOTE: we rely entirely on the schema to make sure that regular snuba
+        # queries are required to send a project_id filter. Some other special
+        # internal query types do not require a project_id filter.
+        project_ids = util.to_list(extension_data['project'])
+        stats.update({'num_projects': len(project_ids)})
+
+        if project_ids:
+            query.add_conditions([('project_id', 'IN', project_ids)])
+
+        turbo = request_settings.get('turbo', False)
+        if not turbo:
+            final, exclude_group_ids = get_projects_query_flags(project_ids)
```

Comment:
How does this work for a dataset that needs the project extension but does not have group_id ?

Reply:
Hmmm that is a good question. Going to think about it more but my initial thoughts are that we will need to create a `ProjectWithGroup` extension processor and only give that to the `Events` dataset.